arch/lkl/kernel/syscalls.c

```diff
@@ -139,6 +145,23 @@ long lkl_syscall(long no, long *params)
 		no, current->comm, test_tsk_thread_flag(task, TIF_HOST_THREAD),
 		test_tsk_thread_flag(task, TIF_SIGPENDING));
 
+	/**
+	 * If there is an ongoing shutdown, we assume that the caller (the idle
+	 * host task) already holds the CPU lock.
+	 */
+	if (lkl_shutdown) {
+		ret = run_syscall(no, params);
+		task_work_run();
+		do_signal(NULL);
+
+		if (no == __NR_reboot) {
+			thread_sched_jb();
+		}
+
+		LKL_TRACE("Shutdown syscall=%li done\n", no);
+		return ret;
+	}
+
```

Comment:
I don't understand this comment or code.  During shutdown, any thread can call `lkl_syscall`.  We haven't done a `switch_to_host_task` to make the current task the idle host task and we haven't checked what the current host thread is to make sure that we're not being invoked from multiple threads.  What am I missing?

Reply:
The assumption is that all userspace threads have been terminated at this point. The idle host task is only invoked when this has happened, and it becomes the only task that can issue syscalls at this point.

Reply:
This isn't enforced anywhere.  We aren't waiting for all other threads to exit and we aren't blocking any other threads from doing `'lkl_syscall`.

Reply:
That is correct. How would we enforce this and where?

Reply:
The simplest thing to do is set the current thread's task to be the idle task (or some other task that we create for this purpose and wake up via a `switch_to_host_task`) and modify `lkl_syscall` to check that the calling thread's task is that task and return -ENOTSUP or similar during shutdown from syscally from any other threads.